src/linalg/arpack.c

```diff
@@ -1374,64 +1379,52 @@ int igraph_arpack_unpack_complex(igraph_matrix_t *vectors, igraph_matrix_t *valu
                      IGRAPH_EINVAL);
     }
 
-    IGRAPH_CHECK(igraph_matrix_resize(vectors, nodes, nev * 2));
-    for (i = nev; i < igraph_matrix_nrow(values); i++) {
+    for (i = no_evs -1; i >= nev; i--) {
         IGRAPH_CHECK(igraph_matrix_remove_row(values, i));
     }
 
-    /* Calculate where to start copying */
-    for (i = 0, j = 0, wh = 0; i < nev; i++) {
-        if (MATRIX(*values, i, 1) == 0) { /* TODO: == 0.0 ???? */
-            /* real */
-            j++;
-        } else {
-            /* complex */
-            if (wh == 0) {
-                j += 2;
-                wh = 1 - wh;
-            }
-        }
-    }
-    j--;
+    IGRAPH_CHECK(igraph_matrix_init(&new_vectors, nodes, nev * 2));
+    IGRAPH_FINALLY(igraph_matrix_destroy, &new_vectors);
 
-    /* if (j>=origcol) { */
-    /*   IGRAPH_WARNING("Too few columns in `vectors', ARPACK results are likely wrong"); */
-    /* } */
-
-    /* We copy the j-th eigenvector to the (k-1)-th and k-th column */
-    k = nev * 2 - 1;
-
-    for (i = nev - 1; i >= 0; i--) {
+    new_vector_pos = 0;
+    vector_pos = 0;
+    for (i = 0; i < nev && vector_pos < igraph_matrix_ncol(vectors); i++) {
         if (MATRIX(*values, i, 1) == 0) {
+            /* Real eigenvalue */
+            memcpy(&MATRIX(new_vectors, 0, new_vector_pos), &MATRIX(*vectors, 0, vector_pos), colsize);
```

Comment:
This assumes that igraph matrices are stored in column-major order. This is true, but if we ever decide to change it later, this part of the code will break. I'd rather use a for loop for copying the elements even if it is less efficient. A good compiler should be able to optimize it away anyway, and even if it doesn't, the performance hit is probably negligible.

We can also decide to add an `igraph_matrix_copy_column()` function in `matrix.pmt` if we really want to; this way the assumption about column-major matrices will be encapsulated in `matrix.pmt`.

Reply:
Good point :) Since column-copying isn't used that much ( I think), and the for loops are fine, I think for now just using a for loop every time is no problem.